lib/puppet/indirector/catalog/compiler.rb

```diff
@@ -138,59 +96,105 @@ def common_checksum_type(agent_checksum_type)
-        resource[:checksum] = checksum_type
-      end
-
-      file = resource.to_ral
-
-      if file.recurse?
-        child_resources = file.recurse_remote_metadata.map do |meta|
-          # Don't create a new resource for the parent directory
-          next if meta.relative_path == "."
+      sources = [resource[:source]].flatten.compact
+      next if sources.empty?
+      next unless sources.all? {|source| source =~ /^puppet:/}
+
+      # both need to handle multiple sources
+      if resource[:recurse] == true || resource[:recurse] == 'true' || resource[:recurse] == 'remote'
+        # Construct a hash mapping sources to arrays (list of files found recursively) of metadata
+        options = {
+          :environment        => catalog.environment_instance,
+          :links              => resource[:links] ? resource[:links].to_sym : :manage,
+          :checksum_type      => resource[:checksum] ? resource[:checksum].to_sym : checksum_type.to_sym,
+          :source_permissions => resource[:source_permissions] ? resource[:source_permissions].to_sym : :ignore,
+          :recurse            => true,
+          :recurselimit       => resource[:recurselimit],
+          :ignore             => resource[:ignore],
+        }
+
+        sources_in_environment = true
+
+        source_to_metadatas = {}
+        sources.each do |source|
+          if list_of_data = Puppet::FileServing::Metadata.indirection.search(source, options)
+            basedir_meta = list_of_data.find {|meta| meta.relative_path == '.'}
+            devfail "FileServing::Metadata search should always return the root search path" if basedir_meta.nil?
+
+            if ! basedir_meta.full_path.start_with? environment_path.to_s
+              # If any source is not in the environment path, skip inlining this resource.
+              sources_in_environment = false
+              break
+            end
 
-          child_resource = Puppet::Resource.new(:file, File.join(file[:path], meta.relative_path))
-          child_resource[:source] = meta.source
-          replace_metadata(child_resource, meta)
+            base_content_uri = get_content_uri(basedir_meta, source, environment_path)
+            list_of_data.each do |metadata|
+              if metadata.relative_path == '.'
+                metadata.content_uri = base_content_uri
+              else
+                metadata.content_uri = "#{base_content_uri}/#{metadata.relative_path}"
+              end
+            end
 
-          # Copy parameters from original parent directory
-          file.original_parameters.each do |param, value|
-            # These should never be passed to our children
-            unless [:parent, :ensure, :recurse, :recurselimit, :target, :alias, :source].include? param
-              child_resource[param] = value.to_s
+            source_to_metadatas[source] = list_of_data
+            # Optimize for returning less data if sourceselect is first
```

Comment:
We may not want to do this, depending on what metadata means when it's added to puppetdb. Since that won't be the next release, we can remove it later if needed.